Add contextual logging for ingress controller.

sawsa307 opened the pull request:
* Initialize a logger instance in cmd/glbc/main.go.
* All ingress controller log messages will be prefixed by `IngressController`. 
  Example:
```
I0111 19:55:55.582574  873925 controller.go:170] "IngressController: Ingress added, enqueuing" ingressKey="default/neg-demo-ing" 
I0111 19:55:55.583044  873925 controller.go:585] "IngressController/preSyncGC: Running preSyncGC" ingressKey="default/neg-demo-ing"
I0111 19:55:55.583458  873925 controller.go:406] "IngressController/SyncBackends: Skip syncing instance groups" ingressKey="default/neg-demo-ing"    
```
* Logging for sub-components like HealthChecker will be added in separate PRs.

sawsa307 commented:
/assign @swetharepakula

sawsa307 commented on pkg/controller/controller.go:
```diff
@@ -440,7 +446,7 @@ func (lbc *LoadBalancerController) syncInstanceGroup(ing *v1.Ingress, ingSvcPort
 	// TODO: Remove this after deprecation
 	if utils.IsGCEMultiClusterIngress(ing) {
-		klog.Warningf("kubemci is used for ingress %v", ing)
+		lbc.logger.Info("kubemci is used for ingress", "ingress", ing)
```
Looks like `Info` to me.

swetharepakula commented:
/ok-to-test

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -449,7 +455,7 @@ func (lbc *LoadBalancerController) syncInstanceGroup(ing *v1.Ingress, ingSvcPort
 			return err
 		}
-		if err = updateAnnotations(lbc.ctx.KubeClient, ing, newAnnotations); err != nil {
+		if err = updateAnnotations(lbc.ctx.KubeClient, ing, newAnnotations, lbc.logger); err != nil {
```
Can we add to the context at the beginning of this function to say this is for syncInstanceGroup

sawsa307 replied:
To confirm, basically we want to also add the name of the caller function as a context before pass it down to the helper function `updateAnnotations`?

swetharepakula replied:
like there needs to be identifying information in the log line so that we are able to track the sync all the way down

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -154,53 +154,56 @@ func NewLoadBalancerController(
 			}
 
-			klog.V(2).Infof("Ingress %v added, enqueuing", common.NamespacedName(addIng))
+			ingLogger.V(2).Info("Ingress added, enqueuing")
```
remove all the custom logging levels

swetharepakula reviewed (commented):
Few quick comments. In the ingress controller, we should probably add a context like ingress key here: https://github.com/kubernetes/ingress-gce/blob/bfa7657e1ccb79d93692fd0e1b55f9d3337c4f95/pkg/controller/controller.go#L610.

And then we should pass that logger into all subsequent functions, even it they are methods on the controller object because we want to keep the context throughout the whole sync

sawsa307 commented:
> Few quick comments. In the ingress controller, we should probably add a context like ingress key here:
> 
> https://github.com/kubernetes/ingress-gce/blob/bfa7657e1ccb79d93692fd0e1b55f9d3337c4f95/pkg/controller/controller.go#L610
> 
> .
> And then we should pass that logger into all subsequent functions, even it they are methods on the controller object because we want to keep the context throughout the whole sync

To make sure I'm understanding this correctly, are you suggesting in this specific function sync(), we should create a logger with additional context(like ingress key here), and all the member functions(like hasSynced, preSyncGC, and etc) will use this new logger because we want to keep this context, as it is operating on a specific ingress throughout this function?

swetharepakula commented:
Yes t

> > Few quick comments. In the ingress controller, we should probably add a context like ingress key here:
> > https://github.com/kubernetes/ingress-gce/blob/bfa7657e1ccb79d93692fd0e1b55f9d3337c4f95/pkg/controller/controller.go#L610
> > 
> > .
> > And then we should pass that logger into all subsequent functions, even it they are methods on the controller object because we want to keep the context throughout the whole sync
> 
> To make sure I'm understanding this correctly, are you suggesting in this specific function sync(), we should create a logger with additional context(like ingress key here), and all the member functions(like hasSynced, preSyncGC, and etc) will use this new logger because we want to keep this context, as it is operating on a specific ingress throughout this function?

Yes, we want to be able to query logs for a single Ingress and receive all the relevant logs.

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -378,6 +387,7 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
 	// as being in an error state in the UI
 	lbc.backendLock.Lock()
 	defer lbc.backendLock.Unlock()
+	ingLogger = ingLogger.WithValues("function", "LoadBalancerController.SyncBackends")
```
nit: ingLogger.WithName("SyncBackends")

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -489,14 +499,15 @@ func (lbc *LoadBalancerController) GCBackends(toKeep []*v1.Ingress) error {
 // SyncLoadBalancer implements Controller.
-func (lbc *LoadBalancerController) SyncLoadBalancer(state interface{}) error {
+func (lbc *LoadBalancerController) SyncLoadBalancer(state interface{}, ingLogger klog.Logger) error {
+	ingLogger = ingLogger.WithValues("function", "LoadBalancerController.SyncLoadBalancer")
```
ingLogger.WithName(SyncLoadBalancer)

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -522,64 +533,67 @@ func (lbc *LoadBalancerController) GCv2LoadBalancer(ing *v1.Ingress, scope meta.
 // PostProcess implements Controller.
-func (lbc *LoadBalancerController) PostProcess(state interface{}) error {
+func (lbc *LoadBalancerController) PostProcess(state interface{}, ingLogger klog.Logger) error {
+	ingLogger = ingLogger.WithValues("function", "LoadBalancerController.PostProcess")
```
ingLogger = ingLogger.WithName("PostProcess")

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -522,64 +533,67 @@ func (lbc *LoadBalancerController) GCv2LoadBalancer(ing *v1.Ingress, scope meta.
 	defer lbc.gcLock.Unlock()
-	klog.V(4).Infof("Running preSyncGC for ingress %q. ", key)
-	defer klog.V(4).Infof("Finish preSyncGC for ingress %q. ", key)
+	ingLogger = ingLogger.WithValues("function", "LoadBalancerController.preSyncGC")
```
ingLogger.WithName(preSyncGC)

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -592,18 +606,20 @@ func (lbc *LoadBalancerController) preSyncGC(key string, scope meta.KeyType, ing
 	defer lbc.gcLock.Unlock()
-	klog.V(4).Infof("Running postSyncGC for ingress %q. ", key)
-	defer klog.V(4).Infof("Finish postSyncGC for ingress %q. ", key)
+	ingLogger = ingLogger.WithValues("function", "LoadBalancerController.postSyncGC")
```
ingLogger=ingLogger.WithName(postSyncGC)

swetharepakula commented on pkg/controller/controller.go:
```diff
@@ -828,10 +845,10 @@ func (lbc *LoadBalancerController) defaultFrontendNamingScheme(ing *v1.Ingress)
-func (lbc *LoadBalancerController) ensureFinalizer(ing *v1.Ingress) (*v1.Ingress, error) {
-	ingKey := common.NamespacedName(ing)
+func (lbc *LoadBalancerController) ensureFinalizer(ing *v1.Ingress, ingLogger klog.Logger) (*v1.Ingress, error) {
+	ingLogger = ingLogger.WithValues("function", "LoadBalancerController.ensureFinalizer")
```
ingLogger=ingLogger.WithName("ensureFinalizer")

swetharepakula reviewed (commented):
Few suggestions on how to add the functions to the logger context

sawsa307 commented:
> Few suggestions on how to add the functions to the logger context

I'm not sure if mixing function name and component name is a good idea, but it should be okay if we are only doing it for ingress controller. I've addressed the comments, PTAL!

swetharepakula reviewed (commented):
/lgtm
/approve
